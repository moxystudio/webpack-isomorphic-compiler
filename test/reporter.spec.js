'use strict';

const WritableStream = require('stream').Writable;
const fs = require('fs');
const createCompiler = require('./util/createCompiler');
const configBasicClient = require('./configs/basic-client');
const configBasicServer = require('./configs/basic-server');
const configSyntaxError = require('./configs/syntax-error');

function createInMemoryOutputStream() {
    let output = '';
    const writableStream = new WritableStream();

    return Object.assign(writableStream, {
        _write(chunk, encoding, callback) {
            output += chunk;
            callback();
        },

        getOutput() {
            return output
            // Replace (xxxms) with (10ms)
            .replace(/\(\d+ms\)/g, '(10ms)')
            // Remove trailing spaces in new lines (webpack toString adds a few..)
            .split('\n')
            .map((line) => line.trimRight())
            .join('\n');
        },
    });
}

describe('reporter', () => {
    afterEach(() => createCompiler.teardown());

    it('should report a successful compilation', () => {
        const compiler = createCompiler(configBasicClient, configBasicServer);
        const outputStream = createInMemoryOutputStream();

        return compiler.run({
            report: { output: outputStream },
        })
        .then(() => {
            expect(outputStream.getOutput()).toMatchSnapshot();
        });
    });

    it('should report a failed compilation', () => {
        const compiler = createCompiler(configBasicClient, configSyntaxError);
        const outputStream = createInMemoryOutputStream();

        return compiler.run({
            report: { output: outputStream },
        })
        .catch(() => {})
        .then(() => {
            expect(outputStream.getOutput()).toMatchSnapshot();
        });
    });

    it('should report a fatal error while compiling', () => {
        const compiler = createCompiler(configBasicClient, configBasicServer);
        const outputStream = createInMemoryOutputStream();
        const contrivedError = new Error('foo');

        compiler.client.webpackCompiler.plugin('before-run', (compiler, callback) => callback(contrivedError));

        return compiler.run({
            report: { output: outputStream },
        })
        .catch(() => {})
        .then(() => {
            expect(outputStream.getOutput()).toMatchSnapshot();
        });
    });

    it('should only display stats in the first compilation if options.stats = \'once\'', (done) => {
        const compiler = createCompiler(configBasicClient, configBasicServer);
        const outputStream = createInMemoryOutputStream();
        let callsCount = 0;

        return compiler.watch({
            report: { stats: 'once', output: outputStream },
        }, () => {
            callsCount += 1;

            if (callsCount === 1) {
                fs.writeFileSync(configBasicClient.entry, fs.readFileSync(configBasicClient.entry));
            } else {
                expect(outputStream.getOutput()).toMatchSnapshot();
                done();
            }
        });
    });

    it('should use options.statsOptions when printing stats', () => {
        const compiler = createCompiler(configBasicClient, configBasicServer);
        const outputStream = createInMemoryOutputStream();

        return compiler.run({
            report: { stats: true, statsOptions: { hash: true }, output: outputStream },
        })
        .then(() => {
            expect(outputStream.getOutput()).toMatchSnapshot();
        });
    });
});
