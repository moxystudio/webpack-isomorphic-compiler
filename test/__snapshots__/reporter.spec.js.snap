// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reporter should only display stats in the first compilation if options.stats = 'once' 1`] = `
"[2m•[22m Compiling...
[32m✓[39m Compilation succeeded [2m(10ms)[22m

    [7m CLIENT                   [27m
    [2m━━━━━━━━━━━━━━━━━━━━━━━━━[22m
    [1mAsset[39m[22m     [1mSize[39m[22m  [1mChunks[39m[22m  [1m[39m[22m           [1m[39m[22m[1mChunk Names[39m[22m
    [1m[32mclient.js[39m[22m  2.69 kB       [1m0[39m[22m  [1m[32m[emitted][39m[22m  main

    [7m SERVER                   [27m
    [2m━━━━━━━━━━━━━━━━━━━━━━━━━[22m
    [1mAsset[39m[22m     [1mSize[39m[22m  [1mChunks[39m[22m  [1m[39m[22m           [1m[39m[22m[1mChunk Names[39m[22m
    [1m[32mserver.js[39m[22m  2.69 kB       [1m0[39m[22m  [1m[32m[emitted][39m[22m  main

[2m•[22m Compiling...
[32m✓[39m Compilation succeeded [2m(10ms)[22m

"
`;

exports[`reporter should report a failed compilation 1`] = `
"[2m•[22m Compiling...
[31m✖[39m Compilation failed

    [1mAsset[39m[22m     [1mSize[39m[22m  [1mChunks[39m[22m  [1m[39m[22m           [1m[39m[22m[1mChunk Names[39m[22m
    [1m[32mserver.js[39m[22m  2.89 kB       [1m0[39m[22m  [1m[32m[emitted][39m[22m  main

    [1m[31mERROR in ./test/configs/files/syntax-error.js
    Module parse failed: /Users/andrecruz/Work/moxy/moxy/webpack-isomorphic-compiler/test/configs/files/syntax-error.js Unexpected token (4:0)
    You may need an appropriate loader to handle this file type.
    |
    | console.log('Hello!'
    | [39m[22m

"
`;

exports[`reporter should report a fatal error while compiling 1`] = `
"[2m•[22m Compiling...
[31m✖[39m Compilation failed

    foo
    [0m[90m-[0m [0m[93mreporter.spec.js[0m[90m:[0m[93m63[0m[37m [0m[37mObject.it[0m
    [0m  [0m  [0m[90m/Users/andrecruz/Work/moxy/moxy/webpack-isomorphic-compiler/test/reporter.spec.js:63:32[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-async.js[0m[90m:[0m[93m42[0m[37m [0m[37mObject.<anonymous>[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/build/jasmine-async.js:42:32[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1984[0m[37m [0m[37mattemptAsync[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1984:24[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1939[0m[37m [0m[37mQueueRunner.run[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1939:9[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1966[0m[37m [0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1966:16[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1909[0m[37m [0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1909:9[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-async.js[0m[90m:[0m[93m68[0m[37m [0m[37mObject.fn[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/build/jasmine-async.js:68:11[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1984[0m[37m [0m[37mattemptAsync[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1984:24[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1939[0m[37m [0m[37mQueueRunner.run[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1939:9[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1927[0m[37m [0m[37mQueueRunner.execute[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1927:10[0m

[2m•[22m Compiling...
[31m✖[39m Compilation failed

    foo
    [0m[90m-[0m [0m[93mreporter.spec.js[0m[90m:[0m[93m63[0m[37m [0m[37mObject.it[0m
    [0m  [0m  [0m[90m/Users/andrecruz/Work/moxy/moxy/webpack-isomorphic-compiler/test/reporter.spec.js:63:32[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-async.js[0m[90m:[0m[93m42[0m[37m [0m[37mObject.<anonymous>[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/build/jasmine-async.js:42:32[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1984[0m[37m [0m[37mattemptAsync[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1984:24[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1939[0m[37m [0m[37mQueueRunner.run[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1939:9[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1966[0m[37m [0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1966:16[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1909[0m[37m [0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1909:9[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-async.js[0m[90m:[0m[93m68[0m[37m [0m[37mObject.fn[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/build/jasmine-async.js:68:11[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1984[0m[37m [0m[37mattemptAsync[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1984:24[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1939[0m[37m [0m[37mQueueRunner.run[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1939:9[0m
    [0m  [0m[90m-[0m [0m[93mjasmine-2.5.2.js[0m[90m:[0m[93m1927[0m[37m [0m[37mQueueRunner.execute[0m
    [0m  [0m  [0m[90m[webpack-isomorphic-compiler]/[jest-jasmine2]/vendor/jasmine-2.5.2.js:1927:10[0m

"
`;

exports[`reporter should report a successful compilation 1`] = `
"[2m•[22m Compiling...
[32m✓[39m Compilation succeeded [2m(10ms)[22m

"
`;

exports[`reporter should use options.statsOptions when printing stats 1`] = `
"[2m•[22m Compiling...
[32m✓[39m Compilation succeeded [2m(10ms)[22m

    [7m CLIENT                   [27m
    [2m━━━━━━━━━━━━━━━━━━━━━━━━━[22m
    Hash: [1ma15ebe812cb1f543fee8[39m[22m
        [1mAsset[39m[22m     [1mSize[39m[22m  [1mChunks[39m[22m  [1m[39m[22m           [1m[39m[22m[1mChunk Names[39m[22m
    [1m[32mclient.js[39m[22m  2.69 kB       [1m0[39m[22m  [1m[32m[emitted][39m[22m  main

    [7m SERVER                   [27m
    [2m━━━━━━━━━━━━━━━━━━━━━━━━━[22m
    Hash: [1mfb1cf31fcb450d9c6374[39m[22m
        [1mAsset[39m[22m     [1mSize[39m[22m  [1mChunks[39m[22m  [1m[39m[22m           [1m[39m[22m[1mChunk Names[39m[22m
    [1m[32mserver.js[39m[22m  2.69 kB       [1m0[39m[22m  [1m[32m[emitted][39m[22m  main

"
`;
